From 4b9f00fe5fc9a65c1ff631bb0b3677775291a4e4 Mon Sep 17 00:00:00 2001
From: Rune Lillesveen <futhark@chromium.org>
Date: Wed, 06 Nov 2024 19:36:44 +0000
Subject: [PATCH] [@container] style() query unset on root element crashes

Discovered when trying to fix issue 376834363.

StyleResolver::ComputeValue() does not set up a parent style for the
root element which meant applying unset would crash. There was already a
check for inherit. Do the same for unset.

Bug: 376834363
Change-Id: I075dd462668436f98a6f7924929f5baba6eeba09
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5999996
Commit-Queue: Rune Lillesveen <futhark@chromium.org>
Reviewed-by: Anders Hartvoll Ruud <andruud@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1379181}
---

--- a/third_party/blink/renderer/core/css/resolver/style_builder.cc
+++ b/third_party/blink/renderer/core/css/resolver/style_builder.cc
@@ -103,8 +103,10 @@
 
   bool is_inherit = value.IsInheritedValue();
   bool is_initial = value.IsInitialValue();
-  if (is_inherit && !state.ParentStyle()) {
+  bool is_unset = value.IsUnsetValue();
+  if ((is_inherit || is_unset) && !state.ParentStyle()) {
     is_inherit = false;
+    is_unset = false;
     is_initial = true;
   }
   DCHECK(!is_inherit || !is_initial);
@@ -113,7 +115,7 @@
   if (is_inherit && !is_inherited_for_unset) {
     state.StyleBuilder().SetHasExplicitInheritance();
     state.ParentStyle()->SetChildHasExplicitInheritance();
-  } else if (value.IsUnsetValue()) {
+  } else if (is_unset) {
     DCHECK(!is_inherit && !is_initial);
     if (is_inherited_for_unset) {
       is_inherit = true;
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/css/css-conditional/container-queries/style-query-unset-on-root.html
@@ -0,0 +1,20 @@
+<!DOCTYPE html>
+<title>CSS Container Queries Test: query unset style for custom property on :root should always match</title>
+<link rel="help" href="https://drafts.csswg.org/css-conditional-5/#style-container">
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="support/cq-testcommon.js"></script>
+<style>
+  :root { container-name: --root; }
+  @container --root style(--foo: unset) {
+    #target { color: green; }
+  }
+</style>
+<div id="target"></div>
+<script>
+  setup(() => assert_implements_style_container_queries());
+
+  test(() => {
+    assert_equals(getComputedStyle(target).color, "rgb(0, 128, 0)");
+  }, "Match style(--foo: unset) on :root element");
+</script>
